src_assets/windows/assets/shaders/directx/include/convert_yuv420_packed_uv_ps_base.hlsl

```diff
@@ -0,0 +1,35 @@
+Texture2D image : register(t0);
+SamplerState def_sampler : register(s0);
+
+cbuffer color_matrix_cbuffer : register(b0) {
+    float4 color_vec_y;
+    float4 color_vec_u;
+    float4 color_vec_v;
+    float2 range_y;
+    float2 range_uv;
+};
+
+#include "include/base_vs_types.hlsl"
+
+float2 main_ps(vertex_t input) : SV_Target
+{
+#if defined(LEFT_SUBSAMPLING)
+    float3 rgb_left = image.Sample(def_sampler, input.tex_right_left_center.xz).rgb;
+    float3 rgb_right = image.Sample(def_sampler, input.tex_right_left_center.yz).rgb;
+    float3 rgb = CONVERT_FUNCTION((rgb_left + rgb_right) * 0.5);
+#elif defined(TOPLEFT_SUBSAMPLING)
+    float3 rgb_top_left = image.Sample(def_sampler, input.tex_right_left_top.xz).rgb;
+    float3 rgb_top_right = image.Sample(def_sampler, input.tex_right_left_top.yz).rgb;
+    float3 rgb_bottom_left = image.Sample(def_sampler, input.tex_right_left_bottom.xz).rgb;
+    float3 rgb_bottom_right = image.Sample(def_sampler, input.tex_right_left_bottom.yz).rgb;
+    float3 rgb = CONVERT_FUNCTION((rgb_top_left + rgb_top_right + rgb_bottom_left + rgb_bottom_right) * 0.25);
```

Comment:
Maybe I'm missing something, but based on [this illustration](https://github.com/FFmpeg/FFmpeg/blob/fa20f5cd9e131f22da06ef57bf5aedd87ff51a90/libavutil/pixfmt.h#L673C1-L697), it seems like we're doing top and center instead of left and top-left.

Reply:
Alright, buckle in.
<img width="469" alt="41" src="https://github.com/LizardByte/Sunshine/assets/61738816/50ddf08d-3508-4318-b1bb-b63ab573c66f">
Let's concentrate on the "Type 0", this the only one we currently use and what ffmpeg calls AVCHROMA_LOC_LEFT
![Untitled](https://github.com/LizardByte/Sunshine/assets/61738816/efc7fbc8-f5d6-4666-b566-f0d72f4db835)
I've numbered the pixels in the source image, the red square is the area which target chroma subsample occupies.

We calculate that chroma subsample using box filter, or in other words just average all source pixels that are "covered" by the target pixel, if coverage is partial - the corresponding pixel weight is reduced.

Going back to our subsample, we need pixels 2,5 with 1 weight, and pixel 1,3,4,6 with 0.5 weights.

This can be done with 6 texture fetches, but we're too smart (for our good) and do it in 2.
If we ask the linear texture sampler for a point between pixels 2,3,5,6, it will give us the average of these 4 pixels with even weights. Same thing can be done with 1,2,4,5.
And now if we sum these two pixels - we get the weights we want, because pixels 2,5 are used twice, just need to divide it by 2.

So our vertex shader generates these in-between texture points. Type 2 or AVCHROMA_LOC_TOPLEFT is done in a similar manner, 9 texture fetches are reduced to 4 optimized ones.

Reply:
This is also why literally any scaling breaks our chroma - that requires box filter with different radius and the whole math needs to be redone,

Reply:
And the last part, which we currently don't do. That linear texture sampling and shader averaging must be done in linear light. Since we currently do it in srgb gamma (unless hdr), we end up with less saturated colors than they should be. But that's harder to notice than outright blocky artifacts we had until recently.

Reply:
Ok, thanks for the detailed explanation.